contracts/DataSourceDelegate.sol

```diff
@@ -1,86 +1,241 @@
 import '@jbx-protocol-v2/contracts/interfaces/IJBSingleTokenPaymentTerminalStore.sol';
-import '@jbx-protocol-v2/contracts/interfaces/IJBFundingCycleStore.sol';
-
-import '@uniswap/v3-periphery/libraries/OracleLibrary.sol';
-
-contract DataSourceDelegate is IJBFundingCycleDataSource, IJBPayDelegate, IJBRedemptionDelegate {
-  address public constant weth = address(0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2);
-  address public constant jbx = address(0x3abf2a4f8452ccc2cf7b4c1e4663147600646f66);
-  address public constant jbxEthPool = address(0x48598ff1cee7b4d31f8f9050c2bbae98e17e6b17);
+import '@jbx-protocol-v2/contracts/interfaces/IJBToken.sol';
+import '@jbx-protocol-v2/contracts/libraries/JBConstants.sol';
+import '@jbx-protocol-v2/contracts/libraries/JBCurrencies.sol';
+import '@jbx-protocol-v2/contracts/libraries/JBFundingCycleMetadataResolver.sol';
+import '@jbx-protocol-v2/contracts/libraries/JBTokens.sol';
+import '@jbx-protocol-v2/contracts/structs/JBFundingCycle.sol';
 
-  uint32 = 120;
+import '@paulrberg/contracts/math/PRBMath.sol';
+import '@uniswap/v3-periphery/contracts/libraries/OracleLibrary.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-core/contracts/interfaces/callback/IUniswapV3SwapCallback.sol';
+
+contract DataSourceDelegate is IJBFundingCycleDataSource, IJBPayDelegate, IUniswapV3SwapCallback {
+  using JBFundingCycleMetadataResolver for JBFundingCycle;
+  event Test(uint256);
+  error unAuth();
+  error Slippage();
+
+  IWETH9 private constant weth = IWETH9(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+  IJBToken private constant jbx = IJBToken(0x3abF2A4f8452cCC2CF7b4C1e4663147600646f66);
+
+  IUniswapV3Pool private constant pool = IUniswapV3Pool(0x48598Ff1Cee7b4d31f8f9050C2bbAE98e17E6b17);
+
+  IJBPayoutRedemptionPaymentTerminal public immutable jbxTerminal;
+  IJBSingleTokenPaymentTerminalStore public immutable terminalStore;
+  IJBDirectory public immutable directory;
+  IJBFundingCycleStore public immutable fundingCycleStore;
+
+  uint32 private constant twapPeriod = 120; // ADD SETTERS, shouldn't be constant + setter for max twap/spot deviation
+  uint256 public constant projectId = 1;
+  uint256 public immutable reservedRate;
+  uint256 private _swapTokenCount;
+  uint256 private _issueTokenCount;
+
+  constructor(IJBPayoutRedemptionPaymentTerminal _jbxTerminal, uint256 _reservedRate) {
+    terminalStore = _jbxTerminal.store();
+    directory = _jbxTerminal.directory();
+    fundingCycleStore = _jbxTerminal.store().fundingCycleStore();
+    jbxTerminal = _jbxTerminal;
+    reservedRate = _reservedRate;
+  }
 
   function payParams(JBPayParamsData calldata _data)
     external
-    view
     override
     returns (
       uint256 weight,
       string memory memo,
       IJBPayDelegate delegate
     )
   {
-    return (0, _data.memo, IJBPayDelegate(address(this)));
+    // The terminal must be the jbx terminal.
```

Comment:
Reset `_swapTokenCount ` and `_issueTokenCount ` both to 0 (and enforce sender to be terminal?), as a user could manually call this when AMM price is high to set `_issueTokenCount ` and then when the price is lower perform the pay and have the terminal set `_swapTokenCount`.  The `didPay(..)` will then perform both methods and the user ends up with double the tokens.

Reply:
Yes, indeed, sender should be terminal only (same as the delegate), nice catch